### PR TITLE
Fix for variable dependency bug when initializing stateful properties.

### PR DESCRIPTION
**Pull request by aeslaughter**
@WilkAndy The first commit should allow you to run your test w/o the "yaqi_hacky" setting.

@idaholab/moose-team The second commit I am attempting to create a test for this fix, but I am unable to reproduce the problem in MOOSE itself. Does anyone have any ideas? Should I even worry about it?

### CODE REVIEW

File: framework/src/base/MooseVariable.C [hunk 1]
```diff
@@ -175,12 +175,6 @@ MooseVariable::prepare()
   _dof_map.dof_indices (_elem, _dof_indices, _var_num);
   _has_nodal_value = false;
   _has_nodal_value_neighbor = false;
-
-  // FIXME: remove this when the Richard's module is migrated to use the new NodalCoupleable interface.
-  if (_dof_indices.size() > 0)
-    _is_defined = true;
-  else
-    _is_defined = false;
```
Comment (andrsd) [inline framework/src/base/MooseVariable.C]: 👍

**Comment by andrsd** [comment]
I want to take a look at the test today...

File: framework/src/base/ComputeMaterialsObjectThread.C [hunk 2]
```diff
@@ -76,6 +77,10 @@ ComputeMaterialsObjectThread::subdomainChanged()
 {
   _need_internal_side_material = _fe_problem.needMaterialOnSide(_subdomain, _tid);
   _fe_problem.subdomainSetup(_subdomain, _tid);
+
+  std::set<MooseVariable *> needed_moose_vars;
+  _materials.updateVariableDependency(needed_moose_vars, _tid);
+  //_fe_problem.setActiveElementalMooseVariables(needed_moose_vars, _tid);
```
Comment (andrsd) [inline framework/src/base/ComputeMaterialsObjectThread.C]: We will want to remove this line in the final version...

Comment (andrsd) [inline framework/src/base/ComputeMaterialsObjectThread.C]: I am wrong here. We need to uncomment the line ;-)

**Comment by andrsd** [comment]
Figured out the test and uncommented the `_fe_problem` line. It all sits in [andrsd/7116](https://github.com/andrsd/moose/tree/7116). You can fetch, merge and interactive autosquash my branch...

**Comment by aeslaughter** [comment]
Thanks @andrsd for figuring out the test for this problem.

@WilkAndy This should fix your problem so you no longer need the "yaqi_hacky" lines.

**Comment by WilkAndy** [comment]
Great, i'll submit my yaqi_hacky removal - #7126 - when this gets merged.  Thanks guys,

a

**Comment by aeslaughter** [comment]
@WilkAndy It might be a while before I get this merged. This issue uncovered a widespread deficiency in how we handle variable dependencies. I have it working for MOOSE but all sorts of modules test fail.

**Comment by WilkAndy** [comment]
That's fantastic really.  i can still work with my yaqi_hacky and we get to make the code more robust

**Comment by WilkAndy** [comment]
Hi @aeslaughter .  When this is nearing completion, could you activate the heavy tests?  i have quite a few that i don't want failing.

**Comment by aeslaughter** [comment]
@bwspenc There is a single modules test failing in this PR, base on my examination it is possible that the gold may be wrong, but I don't know anything about this test, would you be willing to take a look?

In solid mechancs:
```
line_material_symm_tensor_sampler.test...................................................... FAILED (CSVDIFF) 
```

**Comment by dschwen** [comment]
@hchen139 is currently porting this to tensor mechanics and we discussed whether to remove this object and test from solid mechanics entirely.

**Comment by bwspenc** [comment]
@aeslaughter Sure, I can take a look, although I'll have to get up to speed on this yaqi_hacky stuff.  @dschwen is right -- we want to remove this capability.  It was developed before we had vector postprocessors, and we should have removed it long ago. The easiest way to fix this is to delete the test, and deprecate that code if we're OK with having deprecated code be untested.

**Comment by bwspenc** [comment]
I take that back -- I think this is actually a test of the new stuff.  It's something else that we wanted to deprecate.  I'll take a look.

**Comment by dschwen** [comment]
Yeah, the thing that fails is a VPP. It is likely that the TM version will
fail as well (same base class).

**Comment by hchen139** [comment]
@dschwen @bwspenc the failed test in this PR will be fixed in PR #7092, which use the stress_xx auxvariable not previously used in line_material_symm_tensor_test to test the LineMaterialRankTwoSampler VPP

**Comment by andrsd** [comment]
@aeslaughter I rebased this branch on top of current `devel`, it is in [andrsd/pr-7116](https://github.com/andrsd/moose/tree/pr-7116). 2 tests are still failing in modules - but only exodiffs. I do not know if it related or not to these changes...

**Comment by andrsd** [comment]
One of the failing tests is still `line_material_symm_tensor_sampler`... The other one is `generalized_plane_strain_tm_contact.generalized_plane_strain_tm_contact_field`

**Comment by aeslaughter** [comment]
I can take a look again, I forgot all about this branch and it will take me some time to recall what I was doing.

**Comment by aeslaughter** [comment]
@andrsd I was able to re-base this branch and am getting the same failures that you mention, I will see if I can figure them out.

```
solid_mechanics:line_material_symm_tensor_sampler.test...................................... FAILED (CSVDIFF)
tensor_mechanics:line_material_rank_two_sampler.rank_two_sampler............................ FAILED (CSVDIFF)
tensor_mechanics:line_material_rank_two_sampler.rank_two_scalar_sampler..................... FAILED (CSVDIFF)
```

**Comment by aeslaughter** [comment]
3bde916 is the problem with these tests.

```
3bde916b3826269f679c45b8c6f1a8ae64def091 is the first bad commit
commit 3bde916b3826269f679c45b8c6f1a8ae64def091
Author: Andrew E Slaughter <andrew.slaughter@inl.gov>
Date:   Wed Jun 1 15:52:58 2016 -0600

    Update variable dependencies prior to running general userobjects

:040000 040000 1a826b909bbdc8f1d93aae4b4784b8ccddb5b620 8858879c9109f0adfd98761a5de7ec2203b9cf74 M	framework
```

**Comment by aeslaughter** [comment]
I have a patch for Bighorn for these changes.

@YaqiWang @snschune I need some help with YAK/RattleSnake

**Comment by snschune** [comment]
Lol `Output too large. You will need to download the results to see this.` How do I download this?

**Comment by permcody** [comment]
Click on the big blue arrow next to the "Results" header.

**Comment by aeslaughter** [comment]
@permcody I am not sure if this is related to what you had @lindsayad, but I did this work to improve variable dependency a few months ago. Of course a few YAK tests were failing and never resolved so this had been sitting around. I just did a rebase, lets see what happens.

**Comment by lindsayad** [comment]
@aeslaughter Failing tests related to `LineMaterial`...I had the same issue. I don't know if this is the same here, but I had to add `getMatPropDependencies` and `setActiveMaterialProperties` calls before `reinitMaterials` in `LineMaterialSamplerBase::execute` so that my material properties were actually computed when they needed to be. Maybe a similar thing needs to be done for variable dependencies before `reinitElem` is called? This could be an issue for anything that inherits from `GeneralVectorPostprocessor` and `GeneralUserObject` that has variable and/or material dependencies.

**Comment by permcody** [comment]
Yes Alex, it's rare for General objects to access Material properties. Usually this happens because someone needs control over the element or nodal loops so they still want that information. In these few cases, it's reasonable to require that these objects inherit from the MaterialPropertyInterface or otherwise take care of reporting their dependencies to the system.

**Comment by aeslaughter** [comment]
@permcody This interface is currently not using a MooseObject, the deprecated message if new.

https://github.com/idaholab/moose/blob/513f0cb4da8102d80ad00a5474f9f88955ef2152/framework/src/userobject/SideUserObject.C#L34

**Comment by permcody** [comment]
I'm not sure why this PR is open, @aeslaughter? I think this fix has been merged in. At any rate, it's super stale and should probably be closed. @dschwen?

**Comment by aeslaughter** [comment]
This has never been merged, I was never able to get it to pass YAK. This should be re-visisted sometime.

**Comment by snschune** [comment]
The problem might have been solved with the changes for material dependency resolution. I vaguely remember the same set of tests failing: objects using materials that are not executed within threaded loops.

**Comment by dschwen** [comment]
Stale. Feel free to reopen if this is still needed.

**Pull request closed without merging**